064-topic-replication-factor-change.md

```diff
@@ -0,0 +1,291 @@
+For example, the `NotEnoughValidWindowsException` can be raised when Cruise Control has just started and the cluster model is still building, but also when there is a configuration error which prevents broker metrics collection.
+A similar example is the `OptimizationFailureException` raised when some hard goal is violated, which can be temporary if the configured network capacity is less than the real network capacity and there is a traffic peak.
+Instead, the `OngoingExecutionException` is clearly temporary, and it means that a task execution was attempted while there is already one that is being executed by Cruise Control.
+Another temporary example is the `RuntimeException` when Cruise Control's task queue is full, and the Topic Operator periodically retries until there is room for it.
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause disruption to producers with `acks=all`.
+When this happens, the Topic Operator logs a warning but continues with the replicas change, because the KafkaRoller ignores topics with RF < minISR, and they don't even show up as under replicated in Kafka metrics.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+When a managed KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, but the task execution continues in Cruise Control.
+There is little benefit in using the `stop_proposal_execution` endpoint, because most replicas change tasks will have only one batch that can't be stopped.
+No task will remain hanging, the only problem is that Cruise Control does some unnecessary work executing a batch for a topic that has been deleted.
+
+When a manged topic with replicas change is deleted in Kafka, the Topic Operator recreates the topic with the target replication factor.
+
+When Topic Operator is restarted, both pending and ongoing replicas changes are recovered from KafkaTopic status.
+The ongoing state includes the `sessionId`, which is the unique identifier of a Cruise Control task, which includes one or more replicas changes.
+
+When Cruise Control is restarted, all active tasks are lost, as there is no persistent memory.
+The Topic Operator detects this event when the ongoing replicas changes check returns no tasks for a known `sessionId`.
+All pending and ongoing replicas changes are still stored in KafkaTopic status, and new tasks can be created as soon as Cruise Control becomes ready.
+
+## Affected/not affected projects
+
+The following components will be affected by this proposal:
+
+- Topic Operator: this component will drive the whole replication factor change process.
+- Cluster Operator: this component will be responsible to initialize the Cruise Control environment variables.
```

Comment:
didn't we talked about factoring out the Cruise Control API logic in some common place and use it in the UTO as well? Should it be worth mentioning it here?

Reply:
This is declared before:

> The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.

Here I'm declaring components, not modules.

Reply:
Ok ... I didn't give the right weight to that sentence then :-)